### ISSUE TITLE
kubernetes discovery "/apps" endpoint bug or question

### ISSUE THREAD

**Issue by wind57** [spring-cloud/spring-cloud-kubernetes#1528]
As the title says, I am not sure if this should be treated as a bug or this is a question. 

In the kubernetes discovery controller image, we have such an endpoint:

```
	@GetMapping("/apps")
	public Flux<Service> apps() {
		return reactiveDiscoveryClient.getServices().flatMap(service -> reactiveDiscoveryClient.getInstances(service)
				.collectList().flatMap(serviceInstances -> Mono.just(new Service(service, serviceInstances))));
	}
```

see [here](https://github.com/spring-cloud/spring-cloud-kubernetes/blob/main/spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-discoveryserver/src/main/java/org/springframewok/cloud/kubernetes/discoveryserver/DiscoveryServerController.java#L42..#L46)

It returns a `Service`. This is used, in turn, by [discovery client](https://github.com/spring-cloud/spring-cloud-kubernetes/blob/main/spring-cloud-kubernetes-discovery/src/main/java/org/springframework/cloud/kubernetes/discovery/KubernetesDiscoveryClient.java#L70..#L77).

Discovery Client as seen above does not actually care that the return type is a `Service`, as all it gets from that response is the `name` of the service; and that is on par with every other discovery implementation we have. 

---

The problem is that we are doing a lot of computation that clients don't really need:

```
return reactiveDiscoveryClient.getServices().flatMap(service -> reactiveDiscoveryClient.getInstances(service)
				.collectList().flatMap(serviceInstances -> Mono.just(new Service(service, serviceInstances))));
```

specifically: `reactiveDiscoveryClient.getInstances(service)....`, when returning `reactiveDiscoveryClient.getServices()` would have been enough. 

1) If this is treated as a bug, things are easy. "break" compatibility with anyone potentially using `DiscoveryServerController::apps` method, and change it to return `Flux<String>`, basically:

```
	@GetMapping("/apps")
	public Flux<String> apps() {
		return reactiveDiscoveryClient.getServices();
	}
```

I have to admit, this would be at least an interesting change. The server and consumer are separate applications, and users might, for example, upgrade them independently and find out that things broke. But is it worth inflicting such a pain on them? Especially since this is an optimization, not a bug that breaks applications. 

2) Another option is to let this work the way it is for now and do the above change "in the next major release". Document this properly and move on. This is not ideal either, imho. What if someone is already relying on the return type from `/apps`? 

3) What I am really inclined to do is different. Let's introduce a `DiscoveryServerControllerV2` in the discovery server, that will do the proper thing everywhere. In the client, let's make one more additional call to see if such a `V2` endpoint is present:

- if it's present, use `/apps/v2` instead of `/apps` in discovery client
- if it's not, do whatever we are doing now without any changes

**Comment by ryanjbaxter** [comment]
I don’t think clients only care about the service name, there are other things like the service metadata that clients care about

**Comment by ryanjbaxter** [comment]
Ah I see that you are saying now.  The client code you pointed to looks to be the wrong method, but I get it.

Honestly when I created the API I modeled it after what the Neflix Eureka server looked like
https://github.com/Netflix/eureka/wiki/Eureka-REST-operations

So for the Eureka Client I believe we do something similar where we make a call to /apps and just extract the names.

I am not sure its such a big deal that we do that really, but we could just create a new endpoint like /apps/names which just returns a list of names and use that.  We wouldn't break anyone then.
